lib/streamlit/runtime/state/query_params.py

```diff
@@ -0,0 +1,118 @@
+# Copyright (c) Streamlit Inc. (2018-2022) Snowflake Inc. (2022)
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from dataclasses import dataclass, field
+from typing import Dict, Iterator, List, MutableMapping, Union
+
+from streamlit.proto.ForwardMsg_pb2 import ForwardMsg
+
+
+@dataclass
+class QueryParams(MutableMapping[str, str]):
+    """A lightweight wrapper of a dict that sends forwardMsgs when state changes.
+    It stores str keys with str and List[str] values.
+    """
+
+    _query_params: Dict[str, Union[List[str], str]] = field(default_factory=dict)
+
+    def __init__(self):
+        # avoid using ._query_params as that will use __setattr__,
+        # which itself relies on `_query_params` being defined
+        self.__dict__["_query_params"] = {}
+
+    def __iter__(self) -> Iterator[str]:
+        return iter(self._query_params.keys())
+
+    def __getitem__(self, key: str) -> str:
+        try:
+            # avoid using ._query_params as that will use __setattr__,
+            # which itself relies on __getitem__
+            value = self.__dict__["_query_params"][key]
+            if isinstance(value, list):
+                if len(value) == 0:
+                    return ""
+                else:
+                    return value[-1]  # type: ignore
```

Comment:
It's odd to have this `# type: ignore` but because of the `value = self.__dict__["_query_params"][key]`, mypy says value is `Dict[str, Any]` which is not correct.

Reply:
The issue is that `mypy` can't determine what the type of `self.__dict__["_query_params"]` is since it can't infer that it's the same thing as the `_query_params` attr. It sees the line `self.__dict__["_query_params"] = {}` above and infers the type of the dict is `Dict[str, Any]` since you can't assign a type that's any more strict than that from if you can't associate that line with the type annotation for `_query_params`. 

You could fix this by adding a type annotation to the `self.__dict__["_query_params"] = {}` line instead of adding type-ignores, but this issue might go away entirely if we're just remove `__{get/set/del}attr__` methods from this class entirely (see my comment below)